This notebook uses an R kernel.

# Differential Gene Expression using DESeq2

Author: Zhongyi (James) Guo <br>
Date: 10/28/2024

## Import Packages

In [1]:
getwd()

[1] "/Users/zhongyiguo/Desktop/SCA-DL-DGE/script/deseq2"

In [2]:
.libPaths()

[1] "/Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/library"

In [3]:
library(tidyverse)
library(DESeq2)

── Attaching core tidyverse packages ──────────────────────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.0     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter()  masks stats::filter()
✖ purrr::flatten() masks jsonlite::flatten()
✖ dplyr::lag()     masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Warning message:
“package ‘DESeq2’ was built under R version 4.3.3”
Loading required package: S4Vectors

Warning message:
“package ‘S4Vectors’ was built under R version 4.3.2”
Loading required package: stats4

Loading required package: BiocGenerics


Attaching package: ‘BiocGenerics’


The fo

## Import Data

In [4]:
# count matrix
count_clean <- read_tsv('../../result/deseq2//count_clean.tsv') |> 
    as.data.frame() |>
    mutate(across(where(is.double), as.integer))
head(count_clean)

Rows: 58174 Columns: 9
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: "\t"
chr (1): NAME
dbl (8): Sample_1, Sample_2, Sample_3, Sample_4, Sample_5, Sample_6, Sample_...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


NAME            Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 Sample_6
1 ENSG00000223972   13       18       3        1        5       11     
2 ENSG00000227232 1087     1002     182      531      200      114     
3 ENSG00000278267   23       33       1       12        6        1     
4 ENSG00000243485    0        3       0        1        2        2     
5 ENSG00000284332    0        0       0        0        0        0     
6 ENSG00000237613    0        0       0        0        0        0     
  Sample_7 Sample_8
1  17       10     
2 319      172     
3   4        1     
4   0        2     
5   0        0     
6   2        0

In [5]:
# meta data
meta_data <- read_csv('../../result/deseq2/meta_data.csv')
meta_data$status <- as.factor(meta_data$status)
meta_data

Rows: 8 Columns: 2
── Column specification ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr (2): sample, status

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


sample   status
1 Sample_1 IS    
2 Sample_2 IS    
3 Sample_3 IS    
4 Sample_4 IS    
5 Sample_5 C     
6 Sample_6 C     
7 Sample_7 C     
8 Sample_8 C

## DESeq2

In [6]:
dds <- DESeqDataSetFromMatrix(countData = count_clean, 
                              colData = meta_data, 
                              design = ~status, 
                              tidy = TRUE)

In [7]:
dds <- DESeq(dds)

estimating size factors

estimating dispersions

gene-wise dispersion estimates

mean-dispersion relationship

final dispersion estimates

fitting model and testing



In [8]:
res <- results(dds)

In [9]:
sig_gene <- res |>
    as.data.frame() |>
    filter(abs(log2FoldChange) > 1 & padj < 0.05) |>
    rownames_to_column('gene')
sig_gene

gene            baseMean     log2FoldChange lfcSE     stat      
1    ENSG00000227232  420.856227   1.877341      0.4063936  4.619515 
2    ENSG00000278267    8.613466   2.615336      0.8545570  3.060458 
3    ENSG00000268903   18.716412   5.774899      0.9753198  5.921032 
4    ENSG00000269981   12.547480   5.752313      1.2084716  4.759990 
5    ENSG00000239906    3.579557   3.791956      1.4671741  2.584531 
6    ENSG00000279457  551.484304   2.532600      0.3827524  6.616811 
7    ENSG00000250575   53.860707   5.050183      0.6465986  7.810384 
8    ENSG00000225972 5794.202467   2.717834      0.5158510  5.268641 
9    ENSG00000229344 1565.304483   2.235852      0.6209036  3.600965 
10   ENSG00000230092   18.129602  -2.704409      0.6059129 -4.463362 
11   ENSG00000225880   20.472268   1.258665      0.4842828  2.599029 
12   ENSG00000188976 1669.458194   2.215552      0.4661518  4.752854 
13   ENSG00000187961  122.344404   4.472072      0.6209385  7.202117 
14   ENSG00000187583   20.898840   3.961467      0.7004398  5.655685 
15   ENSG00000187642    4.799924   5.178611      1.3523668  3.829294 
16   ENSG00000188290   44.552088   5.900809      0.9147513  6.450725 
17   ENSG00000187608  159.276378   4.757089      0.7536132  6.312374 
18   ENSG00000188157 2087.439527   3.499346      0.4626582  7.563567 
19   ENSG00000217801   15.564120   3.414835      0.9482732  3.601109 
20   ENSG00000131591  131.759397   3.640088      0.4177733  8.713071 
21   ENSG00000186827    4.245152   3.038250      1.2194445  2.491503 
22   ENSG00000078808 3620.721822   3.373225      0.4949208  6.815686 
23   ENSG00000176022  444.546784   1.496889      0.3764022  3.976834 
24   ENSG00000184163    2.893892   5.195532      1.6950624  3.065098 
25   ENSG00000160087  962.164783   1.500618      0.3438446  4.364234 
26   ENSG00000162572   29.098453   3.384583      0.5600258  6.043621 
27   ENSG00000131584  259.077541   3.363116      0.5119039  6.569819 
28   ENSG00000169972  129.551677   4.995285      0.5053625  9.884559 
29   ENSG00000127054 1455.046421   2.190835      0.3287388  6.664365 
30   ENSG00000224051  360.864448   2.834700      0.3671056  7.721757 
⋮    ⋮               ⋮            ⋮              ⋮         ⋮         
8271 ENSG00000224747      3.46859 -3.110887      1.1444715  -2.718187
8272 ENSG00000183250     29.47465  1.718245      0.5200451   3.304031
8273 ENSG00000160256    368.12782  3.054209      0.5849579   5.221246
8274 ENSG00000276529     19.20132  3.288911      0.6519181   5.044975
8275 ENSG00000235374     31.83771  1.672657      0.4508163   3.710284
8276 ENSG00000215447    375.29885 -1.597961      0.3675194  -4.347964
8277 ENSG00000273796     19.50656  2.678288      0.7439840   3.599926
8278 ENSG00000182871   4949.70701  3.963749      0.5313740   7.459434
8279 ENSG00000173638    463.07423  2.592352      0.4721359   5.490690
8280 ENSG00000142156   1229.80650  1.566137      0.6566986   2.384865
8281 ENSG00000142173    187.62531 -2.592745      0.8237193  -3.147608
8282 ENSG00000160285   2393.22180  2.454924      0.4540860   5.406298
8283 ENSG00000160298    163.82180  3.005793      0.4503778   6.673938
8284 ENSG00000230982     45.69507 -1.347455      0.3902285  -3.452990
8285 ENSG00000212932     77.05785  1.457637      0.4337291   3.360708
8286 ENSG00000211459 170110.39591 -4.205954      0.3739236 -11.248164
8287 ENSG00000210077     77.69081 -2.410515      0.5907026  -4.080760
8288 ENSG00000210082 438110.94831 -2.245042      0.3373900  -6.654145
8289 ENSG00000209082    161.58689 -1.643943      0.4025455  -4.083869
8290 ENSG00000210100     58.66870 -2.980580      0.6465707  -4.609829
8291 ENSG00000210117     76.13965 -2.885473      0.6597042  -4.373888
8292 ENSG00000210144   1593.65367 -2.086375      0.3944385  -5.289480
8293 ENSG00000210154     11.41673 -3.165036      1.0329754  -3.064000
8294 ENSG00000210156     96.39139 -2.462599      0.5393741  -4.565660
8295 ENSG00000228253  40187.88859 -1.481266      0.4260559  -3.476694
8296 ENSG000002

In [10]:
dim(sig_gene)

[1] 8300    7

In [11]:
dim(sig_gene)[1] / dim(count_clean)[1]

[1] 0.1426754

14.27% of all genes were differentially expressed between IS and C.

## Save Data

In [12]:
saveRDS(dds, file = "../../result/deseq2/dds.rds")
write_csv(sig_gene, "../../result/deseq2/sig_gene.csv")

## Conclusion

In this notebook, we used DESeq2 to perform differential gene expression analysis, identifying genes with adjusted p-values below 0.05 and an absolute log2 fold change greater than 1, indicating a two-fold expression change between IS and Control conditions. These genes were statistically significant.

In [13]:
sessionInfo()

R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS 15.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Los_Angeles
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] DESeq2_1.42.1               SummarizedExperiment_1.32.0
 [3] Biobase_2.62.0              MatrixGenerics_1.14.0      
 [5] matrixStats_1.3.0           GenomicRanges_1.54.1       
 [7] GenomeInfoDb_1.38.8         IRanges_2.36.0             
 [9] S4Vectors_0.40.2            BiocGenerics_0.48.1        
[11] lubridate_1.9.3             forcats_1.0.0              
[13] stringr_1.5.1               dpl